### PR TITLE
docs: add sizing explanation to bloom filter docs in parquet

### PR DESCRIPTION
**Pull request by hiltontj**
Added documentation detailing the sizing of bloom filters in the parquet crate.

# Which issue does this PR close?

There is currently no issue for this change.

# Rationale for this change
 

There is a storage penalty for using bloom filters in parquet, so I felt having the explanation of how they are sized, based on the chosen FPP and NDV parameters, would be helpful to users.

# What changes are included in this PR?

This only includes documentation changes to the `bloom_filter` module within the `parquet` crate.

# Are there any user-facing changes?

This is a documentation change, so will be user-facing.

### CODE REVIEW

File: parquet/src/bloom_filter/mod.rs [hunk 1]
```diff
@@ -16,7 +16,61 @@
...
+//!
+//! ```text
+//! b = NP2(m/8) / 32
+//! ```
+//!
+//! Where, `NP2` denotes *the next power of two*, and `m` is divided by 8 to be represented as bytes.
+//!
+//! Here is a table of calculated sizes for various FPP and NDV:
+//!
+//! | NDV       | FPP       | Nb      | Size (KB) |
```
Comment (hiltontj) [inline parquet/src/bloom_filter/mod.rs:53]: This should be `b`, not `Nb`

File: parquet/src/bloom_filter/mod.rs [hunk 2]
```diff
@@ -16,7 +16,61 @@
...
+//!
+//! SBBFs use eight hash functions to cleanly fit in SIMD lanes<sup>[2][sbbf-paper]</sup>, therefore
+//! `k` is set to 8. The SBBF will spread those `m` bits accross a set of `b` blocks that
+//! are each 256 bits, i.e., 32 bytes, in size. The number of blocks is chosen as:
+//!
+//! ```text
+//! b = NP2(m/8) / 32
+//! ```
+//!
+//! Where, `NP2` denotes *the next power of two*, and `m` is divided by 8 to be represented as bytes.
```
Comment (hiltontj) [inline parquet/src/bloom_filter/mod.rs:49]: There may be a more idiomatic way of expressing *the next power of two* as a function than `NP2`.

**Comment by hiltontj** [comment]
Looks like Clippy is unhappy after 1.78. I will likely wait for that to get fixed upstream and then merge/rebase.

File: parquet/src/bloom_filter/mod.rs [hunk 3]
```diff
@@ -16,7 +16,61 @@
...
 
 //! Bloom filter implementation specific to Parquet, as described
-//! in the [spec](https://github.com/apache/parquet-format/blob/master/BloomFilter.md).
+//! in the [spec][parquet-bf-spec].
+//!
+//! # Bloom Filter Size
+//!
+//! Parquet uses the [Split Block Bloom Filter][sbbf-paper] (SBBF) as its bloom filter
+//! implementation. For each column upon which bloom filters are enabled, there will be an SBBF
+//! stored in the metadata for each row group in the parquet file. The size of each filter is
```
Comment (alamb) [inline parquet/src/bloom_filter/mod.rs:25]: I think technically the bloom filter is not stored in the metadata, but rather the metadata stores the offset and length of the bloom filter location ([length](https://docs.rs/parquet/latest/parquet/file/metadata/struct.ColumnChunkMetaData.html#method.bloom_filter_length) and [offset](https://docs.rs/parquet/latest/parquet/file/metadata/struct.ColumnChunkMetaData.html#method.bloom_filter_offset)) -- this is important as it means you can read the metadata without the bloom filter information

Perhaps we can change this to something like

```suggestion
//! implementation. For each column upon which bloom filters are enabled, the offset and length of an SBBF
//! is stored in  the metadata for each row group in the parquet file. The size of each filter is
```

File: parquet/src/bloom_filter/mod.rs (hunk 2 above)
Comment (alamb) [reply parquet/src/bloom_filter/mod.rs:49]: 🤷  I don't know of one off the top of my head

**Review by alamb — COMMENTED** [review]
Thank you @hiltontj  -- I pulled it down a rendered it -- looks really nice @hiltontj 

I had a one small correction but otherwise I think this is ready to 🚀 

Screenshot 2024-05-02 at 3 39 42 PM

**Comment by alamb** [comment]
I believe the clippy failure is unrelated (it is due to the new version of rust, nothing in this PR)

**Comment by alamb** [comment]
PR to fix clippy is here: https://github.com/apache/arrow-rs/pull/5710

**Review by alamb — APPROVED** [review]
Looks great to me -- thanks again @hiltontj

**Comment by alamb** [comment]
I merged up from main to get the clippy fixes

The archery integration test failure does not appear to be related: https://github.com/apache/arrow-rs/issues/5719

**Comment by alamb** [comment]
Thanks again @hiltontj !

**Comment by hiltontj** [comment]
Thanks for closing it out @alamb!

**Pull request merged**